src/gleam/http.gleam

```diff
@@ -52,3 +65,170 @@ pub fn method_to_string(method) {
+  let Ok(
+    uri.Uri(host: Some(host), port: port, path: path, query: query, ..),
+  ) = uri.parse(uri_string)
+  Message(
+    head: RequestHead(method, host, port, path, query),
+    headers: [],
+    body: Nil,
+  )
+}
+
+/// Return the uri that a request was sent to.
+///
+pub fn request_uri(request: Request(a), secure: Bool) -> Uri {
+  let scheme = case secure {
+    True -> "https"
+    False -> "http"
+  }
+
+  let Message(head: RequestHead(_, host, port, path, query), ..) = request
+  Uri(Some(scheme), None, Some(host), port, path, query, None)
+}
+
+/// Construct a `Response` type.
+///
+/// The body type of the returned response is `Nil`,
+/// and should be set with a call to `set_body`.
+pub fn response(status: Int) -> Response(Nil) {
+  Message(head: ResponseHead(status), headers: [], body: Nil)
+}
+
+/// Get the method of a request.
+pub fn method(message: Message(RequestHead, body)) -> Method {
+  let Message(RequestHead(method: method, ..), ..) = message
+  method
+}
+
+/// Get the host of a request.
+pub fn host(message: Message(RequestHead, body)) -> String {
+  let Message(RequestHead(host: host, ..), ..) = message
+  host
+}
+
+/// Get the port of a request, if one was explicitly given.
+pub fn port(message: Message(RequestHead, body)) -> Option(Int) {
+  let Message(RequestHead(port: port, ..), ..) = message
+  port
+}
+
+/// Return the none-empty segments of a request path.
+pub fn path_segments(message: Message(RequestHead, body)) -> List(String) {
+  let Message(RequestHead(path: path, ..), ..) = message
+  uri.path_segments(path)
+}
+
+/// Decode the query from a request.
+pub fn get_query(
+  message: Message(RequestHead, body),
+) -> Result(List(tuple(String, String)), Nil) {
+  let Message(RequestHead(query: query_string, ..), ..) = message
+  case query_string {
+    Some(query_string) -> uri.parse_query(query_string)
+    None -> Ok([])
+  }
+}
+
+/// Get the value associated with a header field, if one was set.
+pub fn get_header(message: Message(head, body), key: String) -> Result(String, Nil) {
+  let Message(headers: headers, ..) = message
+  list.key_find(headers, string.lowercase(key))
+}
+
+/// Set the value of a header field.
+pub fn set_header(
+  message: Message(head, body),
+  key: String,
+  value: String,
+) -> Message(head, body) {
+  let Message(head: head, headers: headers, body: body) = message
+  let headers = list.append(headers, [tuple(string.lowercase(key), value)])
+  Message(head: head, headers: headers, body: body)
+}
+
+/// Fetch the HTTP body as a string.
+///
+/// This method returns the body as a `String` for simplicity.
+/// Internally the body of HTTP messages is represented as `StringBuilder`.
+pub fn get_body(message: Message(head, StringBuilder)) -> String {
+  let Message(body: body, ..) = message
+  string_builder.to_string(body)
+}
+
+/// Add the body to a HTTP message.
+pub fn set_body(
+  message: Message(head, Nil),
+  body: String,
+) -> Message(head, StringBuilder) {
```

Comment:
Why do we do this conversion to StringBuilder?

I would have expected the function that puts a string body returns a `Response(String)` and then we have functions that do conversions such as `Response(String)` -> `Response(StringBuilder)`, or just a `map_body` function. That would really cut down the amount of coversion code we have to write to get full coverage of body types + setter/getter functions.

Reply:
I agree with these comments now. get/set body should not do any conversion